tls/extensions/s2n_npn.c

```diff
@@ -79,3 +79,62 @@ const s2n_extension_type s2n_server_npn_extension = {
     .should_send = s2n_server_npn_should_send,
     .if_missing = s2n_extension_noop_if_missing,
 };
+
+bool s2n_npn_encrypted_should_send(struct s2n_connection *conn)
+{
+    return s2n_server_alpn_should_send(conn);
+}
+
+S2N_RESULT s2n_calculate_padding(uint8_t protocol_len, uint8_t *padding_len)
+{
+    RESULT_ENSURE_REF(padding_len);
+    *padding_len = 32 - ((protocol_len + 2) % 32);
+    return S2N_RESULT_OK;
+}
+
+int s2n_npn_encrypted_extension_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{   
+    uint8_t protocol_len = strlen(conn->application_protocol);
+    POSIX_GUARD(s2n_stuffer_write_uint8(out, protocol_len));
+    POSIX_GUARD(s2n_stuffer_write_bytes(out, (uint8_t*) conn->application_protocol, protocol_len));
+    
+    uint8_t padding_len = 0;
+    POSIX_GUARD_RESULT(s2n_calculate_padding(protocol_len, &padding_len));
+    POSIX_GUARD(s2n_stuffer_write_uint8(out, padding_len));
+    for (size_t i = 0; i < padding_len; i++) {
+        POSIX_GUARD(s2n_stuffer_write_uint8(out, 0));
+    }
+
+    return S2N_SUCCESS;
+    
+}
+
+int s2n_npn_encrypted_extension_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{   
+    uint8_t protocol_len = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint8(extension, &protocol_len));
+#if defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtautological-constant-out-of-range-compare"
+#endif
+    POSIX_ENSURE_LT(protocol_len, sizeof(conn->application_protocol));
+#if defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
+
+    uint8_t *protocol = s2n_stuffer_raw_read(extension, protocol_len);
+    POSIX_ENSURE_REF(protocol);
+    POSIX_CHECKED_MEMCPY(conn->application_protocol, protocol, protocol_len);
```

Comment:
Should `s2n_constant_time_copy_or_dont` be used instead?

Reply:
That's a very specific function; it's kind of hard to explain. It's not providing a copy takes the same time independent of length. It's providing constant-time behavior independent of whether or not the copy actually occurred. 

Essentially if we want the copy to occur, the parameter value "dont" is set to false, and if we don't want the copy to occur, "dont" is set to true. And then the function guarantees those cases take the same amount of time. I'd have to go digging to figure out exactly what we use this function for.

Reply:
If we wanted constant time here, we could also do a busy-work copy of the padding. But I really don't think it's necessary. The alpn extension doesn't bother with padding at all, and the client actually sends the options in plain text.